### PR TITLE
Adds support to share a site (blog) via Reader

### PR DESCRIPTION
**Pull request by stevebaranski**
### Description

Proposes a fix for #10898, which requests the ability to share a site (i.e., blog) via the reader. At the moment, users can share a post via the iOS "share sheet".

This change required some consideration for the fact that `ReaderStreamViewController` is used for heterogenous topic types. It leverages the existing `PostSharingController` and captures a new analytics event.

While the code can be reviewed as-is, it cannot be merged until we've secured approval from design (label affixed) and merged a [supporting PR](https://github.com/wordpress-mobile/WordPress-iOS-Shared/pull/174).

To test:

- Checkout the branch and confirm that existing tests pass.
- Review the code change, taking note of the new analytics event captured.
- Open the Reader and navigate to a selection of topics, sites, and posts and observe the behavior.

Update release notes:

- [X] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### Design Review

I took the liberty to expose a similar experience to the one used to share a blog post in the Reader detail screen. The animations below show the feature working in action with the WP share extension.

Some key distinctions:

- We use the "iOS share" Gridicons.
- We pass the following site metadata to the iOS "share sheet" : site title, description & URL. Its usage varies based on the share destination.

iPhone SE - Portrait - Site :_
10898_iphone

_iPad - Landscape - Discover :_
10898_ipad-landscape

### CODE REVIEW

**Comment by iamthomasbishop** [comment]
Thanks for wrangling this, @stevebaranski – it's a nice improvement! Entry point is good, the only thing I might look to improve here is the content that's used in the case of adding to a new post within the app. I would recommend combining the content into 1 link string, instead of a quote. Example:

`[Site Title] – [Site Description]` – all wrapped in a link.

Question: what is copied to my clipboard when I choose the `copy` option? I'd expect just the URL, based on my experience with other similar apps.

Another thing that I wish we could improve but AFAIK is out of our control re: Activity Views – I wish we could make it clearer what is being shared. I'm thinking similar to how on Action Sheets we can add a description at the top of the Sheet – I'm not sure if we have that capability on Activity Views.

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift [hunk 1]
```diff
@@ -16,7 +17,7 @@ import WordPressFlux
 ///   - Row heights are auto-calculated via UITableViewAutomaticDimension and estimated heights
 ///         are cached via willDisplayCell.
 ///
-@objc open class ReaderStreamViewController: UIViewController, UIViewControllerRestoration {
+@objc class ReaderStreamViewController: UIViewController, UIViewControllerRestoration {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift]: Thanks for cleaning up this class!

**Comment by stevebaranski** [comment]
@frosty thanks for the quick turnaround. I'll respond to each of your points in turn.

> I'll defer to @iamthomasbishop as to the formatting of the shared content. 

I am researching answers to the questions @iamthomasbishop posed, and anticipate sharing my findings on Monday. It will be good to see this in the v11.8 release. :)

> My only question, and this certainly isn't a blocker, is if you know what's causing the share icon to shift position between the stream and detail view? When you tap into / out of a post, you can see it move to the side, as the margin changes.

This is a great catch! Upon closer inspection, `ReaderDetailViewController` was not using the Gridicon version of this image. It was using a PDF in the asset catalog. I tidied that up, and if you take the revised branch for a spin, you should see that it looks a lot more consistent. I noted that the alternate image (`icon-posts-share`) was referenced in `PostPost.storyboard`, but immediately transitioned to the Gridicons reference in `PostPostViewController.viewWillAppear(_:)`. So - I was able to remove the image reference and delete it from the asset catalog.

**Comment by stevebaranski** [comment]
Thanks for your input @iamthomasbishop! Following up:

> I would recommend combining the content into 1 link string, instead of a quote. 

I can modify the proposed behavior so that sharing a blog follows this recommendation. Can I confirm that the existing behavior for sharing a blog **post** (what's shown in the proposal) should remain intact?

> what is copied to my clipboard when I choose the copy option?

The proposed approach copies the site URL to the clipboard. We can absolutely preserve that, applying the title/description logic you recommended above for the other activity types.

> I wish we could make it clearer what is being shared. I'm thinking similar to how on Action Sheets we can add a description at the top of the Sheet – I'm not sure if we have that capability on Activity Views.

Unfortunately this is not possible for [`UIActivityViewController`](https://developer.apple.com/documentation/uikit/uiactivityviewcontroller) to present an accompanying title. I actually filed a Radar with Apple for this very issue [over six years ago](https://openradar.appspot.com/radar?id=2444402), and it was closed as a duplicate. The original bug remains open.

**Comment by iamthomasbishop** [comment]
> I can modify the proposed behavior so that sharing a blog follows this recommendation. Can I confirm that the existing behavior for sharing a blog post (what's shown in the proposal) should remain intact?

I think so, let's not touch that right now.

> The proposed approach copies the site URL to the clipboard. We can absolutely preserve that, applying the title/description logic you recommended above for the other activity types.

That sounds right to me, it's what I would expect to happen if I select the `Copy` option from the Activity Sheet. Roll with it 👍 

> I actually filed a Radar with Apple for this very issue over six years ago, and it was closed as a duplicate. The original bug remains open.

Sad panda :(

**Comment by stevebaranski** [comment]
Thanks @iamthomasbishop - I need to revisit one thing before I think we're good to go. Regarding:

> I would recommend combining the content into 1 link string, instead of a quote. Example:
> [Site Title] – [Site Description] – all wrapped in a link.

When you say "all wrapped in a link", do you mean like something akin to an `<a>` tag?

File: WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController+Sharing.swift [hunk 2]
```diff
@@ -38,15 +40,89 @@ extension ReaderStreamViewController {
...
+        } else {
+            value = title
+        }
+
+        return value
+    }
+
+    public func activityViewController(_ activityViewController: UIActivityViewController, dataTypeIdentifierForActivityType activityType: UIActivity.ActivityType?) -> String {
+
+        return kUTTypeURL as String
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController+Sharing.swift]: Do we need to check the `activityType`, as in `itemForActivityType`? Or is URL suitable for all the various formats we're using?

Comment (stevebaranski) [reply WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController+Sharing.swift]: Good question. Fear not! We account for `SharePost.activityType` in [`67d8b29`](https://github.com/wordpress-mobile/WordPress-iOS/pull/10899/commits/67d8b29bbfbea755045760d8e3df3aa2e17f7ae1), .

**Comment by iamthomasbishop** [comment]
@stevebaranski thanks for the update!

> When you say "all wrapped in a link", do you mean like something akin to an <a> tag?

Yea, that's correct.

> Choosing Messages or similar

I would expect only the URL itself to be contained when sharing to Messages, but I don't feel strongly about that – it's just what I've seen with Apollo and other apps when sharing to Messages. This is also usually the same behavior while sharing to Mail.app, fwiw.

**Comment by stevebaranski** [comment]
This builds locally & tests pass.

I believe the CI check failure is related to the one observed in [this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/10872#issuecomment-458696922); proceeding to merge.

**Pull request merged**